qubes/storage/zfs.py

```diff
@@ -0,0 +1,1219 @@
+                           of the zpool.
+                          For encrypted volumes, this will be
+                            (zpool)/encryption/(encpool)
+                          This may be extended by classes inheriting from
+                           this one, but MUST NOT be shrunk.
+           :type zfs_ns: str
+           :param vm_name: The name of the VM this volume belongs to.
+           :param name: The name of the volume, for example
+                        'private' or 'volatile'
+           :param size: The size (in bytes) the qubesd believes this
+                         volume should have.
+                        Before create() has been called, this directs us to
+                        create a zvol of that size.
+                        If instead we are initializing an existing zvol,
+                        this may be incorrect and should probably be checked.
+                        TODO.
+        """
+
+        self.vm_name = vm_name
+        self.name = name
+        assert self.name
+        self.pool = pool # <- ref to parent ZFSQpool
+        self.log = logging.getLogger(
+            "qubes.storage.zfs.ZFSQzvol.{}".format(zfs_ns))
+        self.log.warning('ZFSQzvol kwargs {}'.format(kwargs))
+        self.zfs_ns = zfs_ns # prefix to use before anything
+        assert self.zfs_ns
+        if self.zfs_ns.startswith("-"):
+            raise qubes.storage.StoragePoolException(
+                "Invalid zfs_ns; prefixed with '-'")
+
+        # TODO verify/update this by retrieving ?
+        # if lzc_exists(whatever):
+        #   self._size = lzc_get_props(whatever, "volsize")
+        self._size = size
+
+        self.vid = "/".join([self.zfs_ns,
+                             "vm",
+                             self.vm_name,
+                             self.name])
+        self.import_path = b"/".join([self.zfs_ns.encode(), b"import",
+                                      self.vm_name.encode()])
+        self._vid_import = b"/".join([self.import_path, self.name.encode()])
+
+        self.save_on_stop = False
+        self.snap_on_start = False
+        self.rw = rw
+        super().__init__(name=self.name,
+                         pool=self.pool,
+                         vid=self.vid,
+                         rw=self.rw,
+                         size=self._size)
+
+        self.log.warning("dataset_path assert {!r} {!r} {!r}".format(self.zfs_ns, b"vm", self.vm_name))
+        self.dataset_path = b"/".join(
+            [
+                self.zfs_ns.encode(),
+                b"vm",
+                self.vm_name.encode(),
+            ]
+        )
+        # sanity check:
+        assert self.dataset_path + b"/" + self.name.encode() == self.vid.encode()
+        self._vid_snap = self.vid.encode() + b"@latest"
+
+
+        if self.save_on_stop:
+            raise qubes.storage.StoragePoolException(
+                "TODO save_on_stop not supported"
+            )
```

Comment:
This code is more important than encryption.

Reply:
Unfortunately I didn't (and still don't) understnad the expected semantics of `save_on_stop`. If you can lay it out in a issue on the other repo I'd be happy to discuss and/or implement it.
As for being more important than encryption: That's entirely subjective. I needed (and need) encrypted, incremental backups - that's why I wrote this. I did not (and do not) need snapshots on start/stop, so I didn't implement that.

Reply:
I had not considered the backups, sorry!